solr/core/src/test/org/apache/solr/update/processor/UpsertConditionTest.java

```diff
@@ -658,6 +657,263 @@ public void givenNullify_whenRunning() {
+      docWithoutOptional.setField("other_field", "Orange");
+
+      assertTrue(condition.matches(null, docWithoutOptional));
+      assertThat(condition.run(null, docWithoutOptional), is(UpsertCondition.ActionType.CONCAT_LC));
+      assertThat(docWithoutOptional.getFieldValue("derived_field"), is("yelloworange"));
+    }
+
+    {
+      SolrInputDocument docWithoutRequired = new SolrInputDocument();
+      docWithoutRequired.setField("field", "Yellow");
+
+      assertTrue(condition.matches(null, docWithoutRequired));
+      assertThat(condition.run(null, docWithoutRequired), is(UpsertCondition.ActionType.CONCAT_LC));
+      assertThat(docWithoutRequired.getFieldValue("derived_field"), nullValue());
+    }
+  }
+
+  @Test
+  public void givenConcatWithFallbacks_whenRunning() {
+    NamedList<String> args = namedList(ImmutableListMultimap.of(
+        "must_not", "NEW.derived_field:*",
+        "action", "concat:derived_field:maybe_prefix?,main_field|fallback_field|last_resort,maybe_suffix?"
+    ));
+
+    UpsertCondition condition = UpsertCondition.parse("concat", args);
+
+    {
+      SolrInputDocument mainPresent = new SolrInputDocument();
+      mainPresent.setField("main_field", "Red");
+      mainPresent.setField("last_resort", "Blue");
+      assertTrue(condition.matches(null, mainPresent));
+      assertThat(condition.run(null, mainPresent), is(UpsertCondition.ActionType.CONCAT));
+      assertThat(mainPresent.getFieldValue("derived_field"), is("Red"));
+    }
+
+    {
+      SolrInputDocument mainAbsent = new SolrInputDocument();
+      mainAbsent.setField("fallback_field", "Green");
+      mainAbsent.setField("last_resort", "Blue");
+      assertTrue(condition.matches(null, mainAbsent));
+      assertThat(condition.run(null, mainAbsent), is(UpsertCondition.ActionType.CONCAT));
+      assertThat(mainAbsent.getFieldValue("derived_field"), is("Green"));
+    }
+
+    {
+      SolrInputDocument onlyLast = new SolrInputDocument();
+      onlyLast.setField("last_resort", "Blue");
+      assertTrue(condition.matches(null, onlyLast));
+      assertThat(condition.run(null, onlyLast), is(UpsertCondition.ActionType.CONCAT));
+      assertThat(onlyLast.getFieldValue("derived_field"), is("Blue"));
+    }
+
+    {
+      SolrInputDocument preAndPost = new SolrInputDocument();
+      preAndPost.setField("fallback_field", "Green");
+      preAndPost.setField("maybe_prefix", "pre");
+      preAndPost.setField("maybe_suffix", "post");
+      assertTrue(condition.matches(null, preAndPost));
+      assertThat(condition.run(null, preAndPost), is(UpsertCondition.ActionType.CONCAT));
+      assertThat(preAndPost.getFieldValue("derived_field"), is("preGreenpost"));
+    }
+
+    {
+      SolrInputDocument onlyPreAndPost = new SolrInputDocument();
+      onlyPreAndPost.setField("maybe_prefix", "pre");
+      onlyPreAndPost.setField("maybe_suffix", "post");
+      assertTrue(condition.matches(null, onlyPreAndPost));
```

Comment:
It's a bit subtle/weird that this will count as matching, but won't have actually set `derived_field`.  I guess it might just need calling out a bit more clearly in a comment or something?

I wonder if we need a test to confirm that having multiple rules where the first rule does this behaves as we expect?  e.g. only have optional fields for the first rule, but then confirm it does/doesn't interfere with later rules?

Reply:
I would expect the conditional clauses for using this action would be `should` for all the contributing fields, and `must_not` for the derived field itself

Reply:
I can add some tests for when there's multiple UpsertCondition in the overall configuration, to check that subsequent rules do/don't match depending on the outcome of an earlier concat action.

But the other issue, making the conditions accurately match only when the concat action will actually be able to set the derived field, I don't think we can achieve that with the current condition matching logic: it's not expressive enough to handle nested AND/OR conditions that we'd need. Effectively we'd need the condition matching to do the same old/new fallback logic that the action currently implements.